### PR TITLE
QVAC-21914 fix: clip FA AUTO on non-coopmat GPUs + ggml-opencl submission hardening (Pixel 9 OOM, S25 Adreno abort)

### PR DESCRIPTION
**Pull request by iancris**
## 🎯 What problem does this PR solve?

`runQwen35ImageTileModeTokensTest` (GPU-default mmproj, `image_max_tokens=4096`, `image_tile_mode=disabled` → a **monolithic 16,384-patch ViT encode**, the heaviest vision encode in the suite) crashes two of the three Android Device-Farm flagships (QVAC-21914):

- **Pixel 9 Pro (Mali-G715 / Vulkan): lmkd kill at ~11.9 GB RSS.** The non-coopmat FA hard-disable introduced in #174 replaces AUTO/ENABLED with DISABLED at warmup, which **bypasses the budget-aware AUTO heuristic** (`clip_resolve_flash_attn_type`) built precisely for this case — the forced explicit-attention path materializes an O(n²·n_head) score matrix that cannot fit at 16k patches.
- **Galaxy S25 Ultra (Adreno 830 / OpenCL): driver abort.** The GPU faults near the tail of the giant encode (`Adreno-GSL log_gpu_snapshot` fires **before any decode work** reaches the device), after which Qualcomm's UMD kills the process from `cl_a8x_cmdbuf_mgr_submit_ibs` (`os_exit`) on the next submission — no CL error ever surfaces to the app. Logcat forensics on both crashing runs revised the initial "decode abort" attribution: the 512-token image-chunk **decode is exonerated** (304 clean full-512-row ubatch decodes in the S25 benchmark); only the monolithic encode — 5–8× beyond anything previously run on this backend — triggers the fault.

**Why S25 and not S26:** not a backend split — S26 Ultra (Adreno 840) runs the **same ggml-opencl path** (`Adreno GPU version 840 found keeping OpenCL backend`, 102× `Chosen GPU OpenCL` in its shard log). S26 completes the identical encode in ~29.5 s on a newer Android-16 UMD; S25 (Android-15 / 830 UMD) faulted ~48 s in. The fault is per-submission-magnitude dependent — fix 2 makes S25 pass **without shortening the workload**.

## 📝 How does it solve it?

Two independent fixes (the S25 crash reproduced twice with fix 1 alone — original signature — confirming independence):

1. **`clip.cpp`: downgrade the non-coopmat FA hard-disable to budget-aware AUTO** (`ef71f2b15`). When the backend answers `ggml_backend_supports_efficient_fa == false`, set `CLIP_FLASH_ATTN_TYPE_AUTO` (not DISABLED) and record `clip_ctx::fa_backend_inefficient`, which now also enables the AUTO cutoff default (4096 patches; previously Mali-detection only). Normal images keep the fast explicit path #174 optimized for; huge encodes fall back to memory-frugal scalar FA — slow-but-alive beats fast-but-OOM. Explicit user `DISABLED` and `MTMD_CLIP_AUTO_FA_MIN_KV` honored as before; backends without the query (Metal/CUDA/OpenCL) unchanged.
2. **`ggml-opencl.cpp`: bound driver submissions** (`ee727efe9`, refined in `d1fe2fad0`).
   - `graph_compute` flushes when the estimated **enqueued work** since the last flush exceeds `GGML_OPENCL_FLUSH_WORK_MB` (default 512 MB, `0` disables) — previously entire multi-thousand-node graphs (~48 s of GPU work) were enqueued with zero intra-graph flushes. Gating on work (not a node count) keeps the per-token LLM decode graph submission-free by construction (it never accumulates near the budget), while the 16k-patch encode flushes dozens of times.
   - `ggml_cl_flash_attn` chunks dispatches along q-rows at `GGML_OPENCL_FA_MAX_NQ` rows (default 4096, `0` disables) with a flush between chunks. The split is exact: the kernel resolves its q row relative to the Q/O/mask base byte offsets, `is_causal` is always 0 (masking is explicit — guarded by `GGML_ASSERT`), and alibi/sinks depend only on the head index — **no `.cl` kernel changes**. For `n_q ≤ 4096` the dispatch path is byte-identical to before.

3. **`clip.cpp`: memory-clamp rework + review hardening** (`d1fe2fad0`). The AUTO cutoff is now clamped by **total** memory (stable, session-independent: explicit scratch ≤ total/4); free memory may only *lower* it via a hard-fit check (scratch ≤ free), never the earlier free/2 heuristic that could push normal-size Mali images onto the slow scalar-FA path under momentary memory pressure. No memory info at all fails **safe** at a conservative 2048-patch cap instead of the raw 4096 default. Plus: both OpenCL env knobs resolved once at init with `strtol` (clamp + warn, not silent-disable) and logged; `is_causal` assert; `n_q == 0` guard.

**CPU paths are untouched**: the warmup change is gated on `backend != backend_cpu`, and both hardening changes live inside the OpenCL backend module.

## 🧪 How was it tested?

All on-device validation via the consumer branch `tetherto/qvac` `test/QVAC-21914-tile-mode-fa-fix` (GPU-default mmproj) carrying this branch as a self-contained vcpkg overlay port; local compile checks for both the CPU-only `mtmd` target and `ggml-opencl` (Khronos headers/ICD loader).

**Crash fix — funcShardE per device, vs the crashing baseline ([28865241320](https://github.com/tetherto/qvac/actions/runs/28865241320), stock v9341.1.5):**

| Device | Stock v9341.1.5 | Fix 1 only (2 runs) | Fix 1 + 2 (this PR) |
|---|---|---|---|
| Pixel 9 Pro (Mali) | ❌ lmkd kill @ ~11.9 GB in tile-mode | ✅ 11/11 passed, twice (tile-mode 386 s) | ✅ 11/11 passed |
| Galaxy S25 Ultra (Adreno 830) | ❌ driver abort in tile-mode | ❌ crashed twice (original `cl_a8x` signature; once earlier in MultiInstance — flaky driver heap corruption) | ✅ **funcShardE 3/3 — first-ever full pass** |
| Galaxy S26 Ultra (Adreno 840) | ✅ | ✅ | ✅ |

- **Targeted S25 gate** ([28898009992](https://github.com/tetherto/qvac/actions/runs/28898009992)): tile-mode test alone on an S25 Ultra — PASS in 266 s, `crashed:false`, **zero driver-fault lines in logcat** (no `log_gpu_snapshot`, no `os_exit`, no SIGABRT).
- **Full Android integration** ([28900670167](https://github.com/tetherto/qvac/actions/runs/28900670167)): all 6 shard-runs × Pixel 9 Pro / S25 Ultra / S26 Ultra — **18/18 device-slots green** (includes `runImageMmprojGpuTest` asserting the GPU projector on all devices).

**Performance A/B — VLM benchmark (qwen3.5-q8, cognitive preset ×3 samples, CPU+GPU cells, 30/30 samples per device, 0 errors), post-hardening ([28900672031](https://github.com/tetherto/qvac/actions/runs/28900672031)) vs pre-hardening baselines (runs #252/#253):**

| Metric | pixel9 CPU / GPU | S25 CPU / GPU (OpenCL) | S26 CPU / GPU (OpenCL) |
|---|---|---|---|
| Quality (Overall %) | 31.1 / 40.0 | 31.1 / **31.1 (Δ0)** | 31.1 / **31.1 (Δ0)** |
| mmproj-encode (ms) | 3920 / 4814 | 1465 / 1061 | 1394 / 669 |
| vs pre-hardening | ±2% | GPU/CPU ratio 0.72 vs 0.76 | GPU 669 vs 672 ms |

GPU quality per-task identical to CPU on both Adreno devices; encode/decode within cross-unit noise of the pre-hardening numbers — **no measurable overhead** from the flushes/chunking (benchmark images are ≤ ~3072 patches, so the chunking path only engages on the huge encodes it exists for). _(This A/B ran the `ee727efe9` hardening; the `d1fe2fad0` refinements are behavior-preserving for these shapes — work-budget gating only removes decode-path flushes, the memory clamp only matters at low free memory — and are re-validated by the S25 gate below.)_

**Unit + parity tests** (`d1fe2fad0`, `tests/`):
- `test-clip-fa-cutoff` — pure-CPU coverage of the AUTO budget arithmetic: fast path survives low free memory, the P2 regression guard (1.5 GB free → hard-fit ~2795, not free/2 ~1976), the no-meminfo fail-safe cap, and edge cases. **Passing locally.**
- `test-opencl-fa-chunking` — chunked-vs-CPU numerical parity (`GGML_OPENCL_FA_MAX_NQ=64`) over unchunked / exact-chunk / partial-last-chunk / `n_q==1`, masked and unmasked. Self-skips without an FP16-capable OpenCL device (PoCL lacks FP16), so it runs on Adreno-class hardware / desktop OpenCL — not on the CI Linux host.

**Re-validation of the refined build** (`d1fe2fad0`) — all green, no regression:
- Targeted S25 tile-mode gate ([28924019133](https://github.com/tetherto/qvac/actions/runs/28924019133)): **PASS in 281 s, `crashed:false`, zero driver-fault lines**.
- Full Android integration ([28926370595](https://github.com/tetherto/qvac/actions/runs/28926370595)): **18/18 device-slots green** — funcShardE 3/3 on Pixel 9 Pro + S25 Ultra + S26 Ultra.
- Benchmark ([28926372393](https://github.com/tetherto/qvac/actions/runs/28926372393)): 30/30 per device, GPU quality Δ0, encode/decode within noise of the pre-review-fix build (S26 GPU 619 ms; S25/S26 GPU decode TPS 30.9/33.5 confirms the work-budget flush leaves the decode hot path unaffected).

**Windows DLL link fix** (`dacb473f9`): `test-clip-fa-cutoff` linked an internal `clip_*` symbol absent from the Windows `mtmd.dll` import lib (no `MTMD_API` export) → `LNK2019` on Windows only. Fixed by moving `clip_fa_effective_min_kv` inline into `clip.h` — no export of an internal helper, builds on every platform. Behavior-identical on Android, re-validated end-to-end at the PR tip: S25 gate ([28928563147](https://github.com/tetherto/qvac/actions/runs/28928563147)) tile-mode PASS 264 s / `crashed:false` / 0 driver-fault lines; full Android integration ([28929513759](https://github.com/tetherto/qvac/actions/runs/28929513759)) **18/18 green**; benchmark ([28929516001](https://github.com/tetherto/qvac/actions/runs/28929516001)) 30/30 per device, GPU quality Δ0, encode within noise (S26 GPU 600 ms).

**PR review comments** (`645e684d7`): `parse_env_i64` now warns when it clamps an oversized `GGML_OPENCL_FLUSH_WORK_MB`/`GGML_OPENCL_FA_MAX_NQ` (was silent); `test-clip-fa-cutoff`'s `n_head=0` case reworked to pass `total_mem>0` so the div-by-zero guard is actually exercised. Windows/CPU-only; no Android behavior change.

**Flush-ordering fix** (`39e0a8641`, @gianni-cor): the work-budget flush ran *before* the budget-crossing node was dispatched, so that node began a fresh batch that could reach graph end unflushed — defeating the bound. Moved the check below dispatch (fused-op `continue` chain → `if/else`, single touch point) so the crossing node is in the flushed batch. Behavioral (tighter bound), so re-validated at PR-head `c720b2b6b`: full Android integration ([28938583459](https://github.com/tetherto/qvac/actions/runs/28938583459)) **18/18 green** (funcShardE 3/3 × Pixel 9 / S25 / S26) + benchmark ([28938585754](https://github.com/tetherto/qvac/actions/runs/28938585754)) 30/30 per device, GPU quality Δ0, encode/decode within noise (S26 GPU 565 ms, S25/S26 decode 29.6/34.6 TPS) — no regression from the reordered flush.

**Review cleanups** (`c720b2b6b`): corrected the "submission-free by construction" flush comment to describe the real per-model cadence, added `n_q <= INT32_MAX` assert, normalized ticket tags, and relocated the OpenCL FA-chunking test in CMakeLists. Non-behavioral — covered by the same PR-head `c720b2b6b` on-device runs as the flush-ordering fix above: full Android integration ([28938583459](https://github.com/tetherto/qvac/actions/runs/28938583459)) **18/18 green** (funcShardE 3/3 × Pixel 9 / S25 / S26) + benchmark ([28938585754](https://github.com/tetherto/qvac/actions/runs/28938585754)) 30/30 per device, GPU quality Δ0.

## ⚠️ Breaking changes

None. Fix 1 only changes behavior on backends that affirmatively report inefficient FA (today: Mali via ggml-vulkan); explicit `DISABLED` and the env override are honored, Metal/CUDA/OpenCL projectors unchanged. Fix 2 is OpenCL-module-only, mathematically exact, byte-identical dispatch for `n_q ≤ 4096`, and both knobs can be disabled at runtime (`GGML_OPENCL_FLUSH_WORK_MB=0`, `GGML_OPENCL_FA_MAX_NQ=0`). CPU inference paths are untouched.

## ✅ Consumer overlay validation (Phase A)

All 6 qvac consumers validated against PR head `c720b2b6b` via a self-contained vcpkg overlay port (branch `validate/QVAC-21914-fabric-overlay-pr181`, commit `d4dee909` — no baseline/version changes). `merge-guard / validate-pr` red is expected on a no-PR dispatch branch; the only retried jobs were infra flakes unrelated to this PR (win32 npm tooling setup, a loaded-runner test timeout):

- classification-ggml: https://github.com/tetherto/qvac/actions/runs/28947785997
- embed-llamacpp: https://github.com/tetherto/qvac/actions/runs/28947788176
- llm-llamacpp: https://github.com/tetherto/qvac/actions/runs/28947790751
- ocr-ggml: https://github.com/tetherto/qvac/actions/runs/28947793129
- translation-nmtcpp: https://github.com/tetherto/qvac/actions/runs/28947795569
- vla-ggml: https://github.com/tetherto/qvac/actions/runs/28947798189

### CODE REVIEW

File: ggml/src/ggml-opencl/ggml-opencl.cpp [hunk 1]
```diff
@@ -3654,6 +3663,31 @@ static ggml_backend_opencl_context * ggml_cl2_init(ggml_backend_dev_t dev) {
...
+        }
+        errno = 0;
+        char * end = nullptr;
+        long long v = std::strtoll(env, &end, 10);
+        if (errno != 0 || end == env || *end != '\0' || v < 0) {
+            GGML_LOG_WARN("ggml_opencl: invalid %s='%s', using default %lld\n",
+                          name, env, (long long) defval);
+            return defval;
+        }
+        return v > maxval ? maxval : (int64_t) v;
```
Comment (yingying0906) [inline ggml/src/ggml-opencl/ggml-opencl.cpp]: A valid but out-of-range value clamps to `maxval` silently. Everywhere else this file overrides a user or driver value it logs a `GGML_LOG_WARN`, for example the Adreno buffer-size reduction. Suggest a WARN here too so a clamped `GGML_OPENCL_FLUSH_WORK_MB` / `GGML_OPENCL_FA_MAX_NQ` isn't invisible.

File: tests/test-clip-fa-cutoff.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    expect_range("huge free cannot raise past total clamp",
+                 clip_fa_effective_min_kv(4096, 2 * GB, 64 * GB, 16), 1550, 1700);
+
+    // No memory info at all: fail SAFE at the conservative constant cap
+    // (2048), never the raw 4096 default (~3.2 GB scratch at n_head=16).
+    expect_eq("no meminfo caps at 2048",        clip_fa_effective_min_kv(4096, 0, 0, 16), 2048);
+    expect_eq("no meminfo keeps smaller cutoff", clip_fa_effective_min_kv(1000, 0, 0, 16), 1000);
+
+    // Degenerate n_head is guarded (treated as 1), not a crash/div-by-zero.
+    expect_eq("n_head=0 guarded", clip_fa_effective_min_kv(2048, 0, 0, 0), 2048);
```
Comment (yingying0906) [inline tests/test-clip-fa-cutoff.cpp]: This case passes `total_mem == 0 && free_mem == 0`, which short-circuits to the NO_MEMINFO cap before either `sqrt(.../n_head)` branch runs, so the `n_head > 0 ? n_head : 1` guard is never hit. The test passes identically even if the guard is deleted. To actually cover it, pass `total_mem > 0` or `free_mem > 0` together with `n_head = 0`.

File: ggml/src/ggml-opencl/ggml-opencl.cpp (hunk 1 above)
Comment (iancris) [reply ggml/src/ggml-opencl/ggml-opencl.cpp]: Good catch — fixed in 645e684d7. The clamp now emits a `GGML_LOG_WARN` ("exceeds max, clamping to N") so an out-of-range value isn't silently trimmed, matching the file's convention for the Adreno buffer-size reductions.

File: tests/test-clip-fa-cutoff.cpp (hunk 2 above)
Comment (iancris) [reply tests/test-clip-fa-cutoff.cpp]: Correct — the case short-circuited on the NO_MEMINFO path before any sqrt(.../n_head) branch ran, so the guard was dead in that assertion. Fixed in 645e684d7: it now passes `total_mem = 16 GB` with `n_head = 0`, so the total-memory clamp actually divides by the guarded head count — without the `n_head > 0 ? n_head : 1` guard the `(int)sqrt(x/0)` path would fail the assertion. Still passes locally.

**Review by yingying0906 — APPROVED** [review]

File: ggml/src/ggml-opencl/ggml-opencl.cpp [hunk 3]
```diff
@@ -4250,6 +4319,18 @@ static ggml_status ggml_backend_opencl_graph_compute(ggml_backend_t backend, ggm
...
         }
 
+        // Single touch point for the submission bound (see the QVAC-21914 note
+        // above): estimate this node's work before dispatching it and submit
+        // the accumulated batch once the budget is exceeded. Fused ops are
+        // accounted by their anchor node — precision is irrelevant here.
+        if (backend_ctx->flush_work_budget > 0) {
+            work_since_flush += ggml_opencl_node_work_estimate(node);
+            if (work_since_flush >= backend_ctx->flush_work_budget) {
+                CL_CHECK(clFlush(backend_ctx->queue));
```
Comment (gianni-cor) [inline ggml/src/ggml-opencl/ggml-opencl.cpp]: Looks like this flush can happen before the node that crosses the budget is enqueued. `work_since_flush` includes the current node, but the dispatch happens below, so this `clFlush` only submits the previous batch and then resets the counter. If a large op, or the last large segment of the graph, is the one that crosses the threshold, it starts a fresh unflushed batch and can still be submitted unbounded at graph completion. I think the budget check needs to run after the fused/non-fused dispatch, or otherwise flush after enqueueing the node that caused the budget to be exceeded.

Comment (iancris) [reply ggml/src/ggml-opencl/ggml-opencl.cpp]: Good catch, you're right — the flush ran before the node was dispatched, so the crossing node reset the counter and began a fresh batch that could reach graph end unflushed. Fixed in 39e0a8641: the budget check now runs *after* the dispatch (I converted the fused-op `continue` chain to `if/else` so every path reaches one flush touch point), so the node that crosses the budget is included in the flushed batch. A single op larger than the budget is still one submission by nature, but it's now flushed immediately rather than deferred to the end-of-graph finish.

**Review by gianni-cor — APPROVED** [review]
